### PR TITLE
UPSTREAM: 00000: debug wedged openshift api cert reload

### PR DESCRIPTION
**Pull request by mfojtik**
This PR will print out details about certificate that was dynamically reloaded. It should help debug wedged openshift api server.

### CODE REVIEW

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/certs/serving.go [hunk 1]
```diff
@@ -57,6 +58,18 @@ type certKeyFileContent struct {
 	Key  []byte
 }
 
+func (c certKeyFileContent) CertSHA256() string {
+	h := sha256.New()
+	h.Write(c.Cert)
+	return fmt.Sprintf("%x", h.Sum(nil))
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/certs/serving.go]: you want this output consistent with output from must-gather inspect-certs so you can easily compare

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/certs/single_certkey_pair.go [hunk 2]
```diff
@@ -97,6 +97,8 @@ func (c *DynamicCertKeyPairLoader) CheckCerts() error {
 	c.currentValue.Store(newRuntimeConfig)
 	c.currentContent = newContent // this is single threaded, so we have no locking issue
 
+	glog.V(2).Infof("Detected certificate change: %q: %s, %q: %s", c.fileReference.Cert, newContent.CertSHA256(), c.fileReference.Key, newContent.KeySHA256())
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/certs/single_certkey_pair.go]: I think this is worthy of info.  But you should make the output match must-gather.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go [hunk 3]
```diff
@@ -227,12 +232,16 @@ func (r *proxyHandler) updateAPIService(apiService *apiregistrationapi.APIServic
...
+
 	if r.proxyTransport != nil && r.proxyTransport.DialContext != nil {
 		newInfo.restConfig.Dial = r.proxyTransport.DialContext
 	}
 	newInfo.proxyRoundTripper, newInfo.transportBuildingError = restclient.TransportFor(newInfo.restConfig)
 	if newInfo.transportBuildingError != nil {
 		glog.Warning(newInfo.transportBuildingError.Error())
 	}
+
+	glog.V(2).Infof("APIService %q: TLS cert=%s, key=%s", newInfo.name, sha256(newInfo.restConfig.CertData), sha256(newInfo.restConfig.KeyData))
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go]: info worthy, but same comment about format.

**Comment by deads2k** [comment]
/hold

aren't there more files to note refreshs in that dynamic package.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/certs/clientca.go [hunk 4]
```diff
@@ -88,8 +88,9 @@ func (c *DynamicCA) CheckCerts() error {
 		return fmt.Errorf("unable to load client CA file %q: %v", c.caFile.Cert, err)
 	}
 	pool := x509.NewCertPool()
-	for _, cert := range certs {
-		pool.AddCert(cert)
+	for i, crt := range certs {
+		pool.AddCert(crt)
+		glog.V(2).Infof("[%d] %q client-ca certificate: %s", i, c.caFile.Cert, getCertDetail(crt))
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/certs/clientca.go]: how often does this happen?

Comment (mfojtik) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/certs/clientca.go]: every time the cert change

**Comment by deads2k** [comment]
known aws flake.  needed to debug cert issues

**Pull request merged**